nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoCsrfFilter.java

```diff
@@ -0,0 +1,153 @@
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.platform.web.common.vh.VirtualHostHelper;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.services.config.ConfigurationService;
+
+/**
+ * Nuxeo CSRF filter, preventing CSRF attacks by rejecting dubious requests.
+ *
+ * @since 10.1
+ */
+public class NuxeoCsrfFilter implements Filter {
+
+    private static final Log log = LogFactory.getLog(NuxeoCsrfFilter.class);
+
+    /**
+     * This configuration property can be set to "true" to disable CSRF protection.
+     */
+    public static final String CSRF_PROTECTION_DISABLED_PROP = "nuxeo.csrf.protection.disabled";
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        // nothing to do
+    }
+
+    @Override
+    public void destroy() {
+        // nothing to do
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+        ConfigurationService configurationService = Framework.getService(ConfigurationService.class);
+        if (configurationService != null && configurationService.isBooleanPropertyTrue(CSRF_PROTECTION_DISABLED_PROP)) {
+            // disabled by configuration
+            chain.doFilter(request, response);
+            return;
+        }
+
+        String method = request.getMethod();
+        if ("GET".equals(method) || "HEAD".equals(method) || "OPTIONS".equals(method) || "TRACE".equals(method)) {
+            // safe method according to RFC 7231 4.2.1
+            chain.doFilter(request, response);
+            return;
+        }
+
+        URI sourceURI = getSourceURI(request);
+        URI targetURI = getTargetURI(request);
+        if (sourceAndTargetMatch(sourceURI, targetURI)) {
+            if (targetURI == null) {
+                // misconfigured server or proxy headers
+                log.error("Cannot determine target URL for CSRF check");
+            }
+            // source and target match, or not provided
+            chain.doFilter(request, response);
+            return;
+        }
+
+        // cross-site request forgery attempt
+        String message = "CSRF check failure";
+        log.warn(message + ": source: " + sourceURI + " does not match target: " + targetURI);
+        response.sendError(HttpServletResponse.SC_FORBIDDEN, message);
+    }
+
+    /** Gets the source URI: the URI of the page from which the request is actually coming. */
+    public URI getSourceURI(HttpServletRequest request) {
+        String source = request.getHeader("Origin");
```

Comment:
Constants again. I see some in your test code.

Reply:
Use `com.google.common.net.HttpHeaders.ORIGIN`